java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/JdbcToArrowUtils.java

```diff
@@ -0,0 +1,343 @@
+                    break;
+                case Types.TIMESTAMP:
+                    // timezone is null
+                    fields.add(new Field(columnName, FieldType.nullable(new ArrowType.Timestamp(TimeUnit.MILLISECOND, null)), null));
+                    break;
+                case Types.BINARY:
+                case Types.VARBINARY:
+                case Types.LONGVARBINARY:
+                    fields.add(new Field(columnName, FieldType.nullable(new ArrowType.Binary()), null));
+                    break;
+                case Types.ARRAY:
+                    // not handled
+//                    fields.add(new Field("list", FieldType.nullable(new ArrowType.List()), null));
+                   break;
+                case Types.CLOB:
+                    fields.add(new Field(columnName, FieldType.nullable(new ArrowType.Utf8()), null));
+                    break;
+                case Types.BLOB:
+                    fields.add(new Field(columnName, FieldType.nullable(new ArrowType.Binary()), null));
+                    break;
+
+                default:
+                    // no-op
+                    break;
+            }
+        }
+
+        return new Schema(fields, null);
+    }
+
+    private static void allocateVectors(VectorSchemaRoot root, int size) {
+        List<FieldVector> vectors = root.getFieldVectors();
+        for (FieldVector fieldVector: vectors) {
+            if (fieldVector instanceof BaseFixedWidthVector) {
+                ((BaseFixedWidthVector) fieldVector).allocateNew(size);
+            } else {
+                fieldVector.allocateNew();
+            }
+            fieldVector.setInitialCapacity(size);
+        }
+    }
+
+    /**
+     * Iterate the given JDBC {@link ResultSet} object to fetch the data and transpose it to populate
+     * the given Arrow Vector objects.
+     *
+     * @param rs ResultSet to use to fetch the data from underlying database
+     * @param root Arrow {@link VectorSchemaRoot} object to populate
+     * @throws Exception
+     */
+    public static void jdbcToArrowVectors(ResultSet rs, VectorSchemaRoot root) throws Exception {
+
+        assert rs != null;
+        assert root != null;
+
+        ResultSetMetaData rsmd = rs.getMetaData();
+        int columnCount = rsmd.getColumnCount();
+
+        allocateVectors(root, DEFAULT_BUFFER_SIZE);
+
+        int rowCount = 0;
+        while (rs.next()) {
+            // for each column get the value based on the type
+
+            // need to change this to build Java lists and then build Arrow vectors
+            for (int i = 1; i <= columnCount; i++) {
+                String columnName = rsmd.getColumnName(i);
+                switch (rsmd.getColumnType(i)) {
+                    case Types.BOOLEAN:
+                    case Types.BIT:
+                        BitVector bitVector = (BitVector) root.getVector(columnName);
+                        bitVector.setSafe(rowCount, rs.getBoolean(i)? 1: 0);
```

Comment:
what about null values? (current code will throw NPE I guess...)

Reply:
In this case, which variable you are referring that can be 'null'. There are null checks before. Can you please elaborate?